numpy/core/src/multiarray/buffer.c

```diff
@@ -538,9 +577,10 @@ _buffer_info_cmp(_buffer_info_t *a, _buffer_info_t *b)
     Py_ssize_t c;
     int k;
 
-    c = strcmp(a->format, b->format);
-    if (c != 0) return c;
-
+    if (a->format != NULL && b->format != NULL) {
+        c = strcmp(a->format, b->format);
+        if (c != 0) return c;
+    }
```

Comment:
Hmm, this looks risky - are you sure it shouldn't be:
```suggestion
    /* null format sorts before empty string */
    c = (a->format != NULL) - (b->format != NULL);
    if (c != 0) return c;
    
    if (a->format != NULL && b->format != NULL) {
        c = strcmp(a->format, b->format);
        if (c != 0) return c;
    }
```
Otherwise `NULL` is considered equal to arbitrary strings, and equality is not transitive

Reply:
Yes, should be correct. if the format is `NULL` it seems OK to replace it with any other format. Now if the old (first) format is `NULL`, we will replace the `NULL` with the actual (second) format.  If the second format is `NULL`, format is ignored completely, so it is fine as well.

Reply:
An empty string would indicate that the itemsize is 0 in the exported buffer, I guess. A size change could seem problematic, but I do not think so.

In theory, changing the itemsize might be dangerous, but it is explicitly stored in the exported buffer info, so while it could be dangerous from a buffer user perspective, I do not think it is dangerous for having incorrect information inside the exported buffer information.

Reply:
Where does this replacing of `NULL` happen?

Reply:
https://github.com/numpy/numpy/pull/17295/files#diff-761c5d4c611d2cd411341182fcee883dR662